File: AtCoder/ABC378/F_Add_One_Edge_2.py
"""
對於一個度數皆為 3 的連通分量，可以計算他相鄰的點中度數為 2 的點的數量 cnt
在這些度數為 2 的點中，任選兩個配對，即可滿足題目要求，故這個連通分量的貢獻為 cnt * (cnt - 1) // 2
可以用 BFS 或 DFS 找到所有度數皆為 3 的連通分量
"""
import sys
sys.setrecursionlimit(10**6)

N = int(input())
edges = [list(map(int, input().split())) for _ in range(N-1)]

g = [[] for _ in range(N + 1)]
deg = [0] * (N + 1)
for u, v in edges:
    g[u].append(v)
    g[v].append(u)
    deg[u] += 1
    deg[v] += 1

vis = [False] * (N + 1)
def dfs(u):
    if deg[u] != 3 or vis[u]:
        return 0
    vis[u] = True
    cnt = 0
    for v in g[u]:
        if deg[v] == 2:
            cnt += 1
        elif deg[v] == 3 and not vis[v]:
            cnt += dfs(v)
    return cnt

ans = 0
for x in range(1, N + 1):
    if deg[x] != 3 or vis[x]:
        continue
    cnt = dfs(x)
    ans += cnt * (cnt - 1) // 2
print(ans)

File: AtCoder/ABC/ABC378/F_Add_One_Edge_2.cpp
/*
對於一個度數皆為 3 的連通分量，可以計算他相鄰的點中度數為 2 的點的數量 cnt
在這些度數為 2 的點中，任選兩個配對，即可滿足題目要求，故這個連通分量的貢獻為 cnt * (cnt - 1) // 2
可以用 BFS 或 DFS 找到所有度數皆為 3 的連通分量
*/

#include <bits/stdc++.h>
using namespace std;
using LL = long long;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int N; cin >> N;
    vector<pair<int, int>> edges(N - 1);
    for (int i = 0; i < N - 1; i++) {
        cin >> edges[i].first >> edges[i].second;
    }
    vector<vector<int>> g(N + 1);
    vector<int> deg(N + 1);
    for (auto [u, v] : edges) {
        g[u].push_back(v);
        g[v].push_back(u);
        deg[u] += 1;
        deg[v] += 1;
    }
    vector<bool> vis(N + 1);
    auto dfs = [&](auto dfs, int u) -> int {
        if (deg[u] != 3 || vis[u]) return 0;
        vis[u] = true;
        int cnt = 0;
        for (int v : g[u]) {
            if (deg[v] == 2) cnt += 1;
            else if (deg[v] == 3 && !vis[v]) cnt += dfs(dfs, v);
        }
        return cnt;
    };
    LL ans = 0;
    for (int i = 1; i <= N; i++) {
        if (deg[i] != 3 || vis[i]) continue;
        int cnt = dfs(dfs, i);
        ans += cnt * (cnt - 1LL) / 2;
    }
    cout << ans << endl;
    return 0;
}